### ISSUE TITLE
AOTAutograd cache always misses even on simple program

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#141790]
### 🐛 Describe the bug

```
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ TORCHINDUCTOR_AUTOGRAD_CACHE=1 TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache python aaa.py 
I1128 17:54:16.961000 1579551 torch/_functorch/_aot_autograd/autograd_cache.py:665] [0/0] AOTAutograd cache miss for key a3fjisonhscsotmkqcifvx22whts72lyron5lrhskrbowx3hfbav
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ TORCHINDUCTOR_AUTOGRAD_CACHE=1 TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache python aaa.py 
I1128 17:54:26.680000 1582644 torch/_functorch/_aot_autograd/autograd_cache.py:665] [0/0] AOTAutograd cache miss for key a3fjisonhscsotmkqcifvx22whts72lyron5lrhskrbowx3hfbav
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ vim aaa.py
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ cat aaa.py
import torch
import torch._dynamo.config



# Artificially generate lots of small kernels
torch._inductor.config.realize_reads_threshold = 1
torch._inductor.config.realize_opcount_threshold = 1
torch._inductor.config.max_fusion_size = 1


N = 2


@torch.compile(fullgraph=True)
def f(a, b):
    for i in range(N):
        a = a + b * i
    return a


f(torch.randn(2, device="cuda", requires_grad=True), torch.randn(2, device="cuda")).sum().backward()
```

### Versions

main

**Comment by ezyang** [comment]
cc @jamesjwu

**Comment by jamesjwu** [comment]
Hmm, the reason this is happening is because AOTAutograd thinks the backward of your graph returns None, so the backward never gets set or compiled, so we never save a cache entry.

```
I1128 19:28:39.666000 592894 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]  ===== Backward graph 0 =====
I1128 19:28:39.666000 592894 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]  <eval_with_key>.3 class GraphModule(torch.nn.Module):
I1128 19:28:39.666000 592894 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]     def forward(self, tangents_1: "f32[2][1]cuda:0"):
I1128 19:28:39.666000 592894 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]         return (None, tangents_1)
```

Basically, I think it thinks a has no derivative? 

If you change the input model to be a = 2 * a + b * i, and there's actually a backward to call, then the backward doesn't return None:

```
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs] TRACED GRAPH
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]  ===== Backward graph 0 =====
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]  <eval_with_key>.3 class GraphModule(torch.nn.Module):
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]     def forward(self, tangents_1: "f32[2][1]cuda:0"):
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]          # File: /home/jjwu/aaa.py:17 in f, code: a = 2*a + b * i
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]         mul_4: "f32[2][1]cuda:0" = torch.ops.aten.mul.Tensor(tangents_1, 2);  tangents_1 = None
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]         mul_5: "f32[2][1]cuda:0" = torch.ops.aten.mul.Tensor(mul_4, 2);  mul_4 = None
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]         return (mul_5, None)
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]
I1128 19:32:54.291000 636032 torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:580] [0/0] [__aot_graphs]
tensor([ 1.4102, -0.6245], device='cuda:0', grad_fn=<CompiledFunctionBackward>)
I1128 19:32:56.595000 636032 torch/_functorch/_aot_autograd/autograd_cache.py:846] [0/0] Writing AOTAutograd cache entry to /tmp/torchinductor_jjwu/aotautograd/aftct3m7parccglz7ecn674wwcpsktbl2cryg2pgwsnpjinovq37/entry
```

I don't really understand why the first model has a backward that returns None, but because of that it's never getting called so we never save the entry. 

If we can pre-detect this, we should be able to save the backward early instead.

**Comment by jamesjwu** [comment]
Ah now that I'm starting at it, is it because from a's perspective you're always adding a constant to it, so a has no gradient?

**Comment by ezyang** [comment]
My hot take is aot autograd cache should hit even for inference only case

**Comment by jamesjwu** [comment]
It does hit for inference only: the only time we can't hit is if you use it for autograd but then never call the backward. (We should definitely handle this particular example, though, where you *do* call the backward, it just seems to be hitting a degenerate case).

Because we have lazy backward compilation, we don't have everything ready by forward compilation. I wanted to try to save a "partial" cache entry that only contains a forward, but we currently don't have a way to serialize the backward module for lazy backward compilation. I also think that adding that complexity seems not worth considering the set of cases that would use it. 

We do want to figure that out, especially since the compiled autograd case will need it, but it's non trivial.

**Comment by ezyang** [comment]
IIRC, I changed things so that we attempt to eagerly compile backwards during forwards (and only delay if that compilation fails). If that happens, it seems intuitively we should be able to cache here.

**Comment by jamesjwu** [comment]
Hmm yes if we eagerly compile the backward we should always be able to cache hit. If you can link an exact diff and input where we eagerly compile the backwards and also don't save to the cache I can investigate

**Comment by ezyang** [comment]
Err, sorry, specifically, the default behavior should already be this (but I also admit I did not check if eager backwards compile actually was happening here)

**Comment by bdhirsh** [comment]
> IIRC, I changed things so that we attempt to eagerly compile backwards during forwards (and only delay if that compilation fails). If that happens, it seems intuitively we should be able to cache here.

The current state as I understand it is that we only do this if we've detected that we have to save any symints for backward (because if inductor wants to guard on any shapes during backward compilation, it's more convenient to do so with dynamo in the loop so dynamo can handle generating shape guards).

Although there are a few downsides to making backward compilation early:

- @BoyuanFeng is forced to turn the donated buffer optimization in inductor when we compile the backward ahead-of-time, since we don't know until runtime whether or not the user will run the backward with `retain_graph=True` (which requires keeping activations around, so they are not safe to donate to inductor

- when compiled autograd is on, we always delay backward compilation. So today we end up compiling the backward ahead of time only to throw it out when compiled autograd is running

- compile time metrics are more annoying to figure out when we sometimes-compile-aot and sometimes-delay-compile

**Comment by bdhirsh** [comment]
But for this specific case, it feels like either:

(1) if you are doing inference, you should probably be using `no_grad()` anyway, since that will make compile emit better code (it won't have to return activations as extra outputs in the forward graph). We will also properly cache hit

(2) or we could try to make this work and figure out how to eagerly save the forward graph to the cache, and only lazily safe the backward to the cache when backward compilation happens

**Comment by ezyang** [comment]
OK, I am convinced there is a legitimate problem here

**Comment by jamesjwu** [comment]
Catching up here: just to be clear, I think the problem here is AOTDispatch thinks this graph may need a backward, and decides to do lazy backward compilation. @ezyang is arguing that we should default to eager backward compilation instead of defaulting to lazy backward compilation, and only lazily do it when we need to, whereas the prevailing thought is the opposite (i.e. do lazy as much as possible). 

From a cache perspective, I am fine with either approach. But if there's any chance we need a backwards compiled it needs to be fully compiled before the cache can save.

Is that a correct summary of this issue? If so I'll update the title and stuff

**Comment by bdhirsh** [comment]
I think that's mostly right. Although would you agree that it's possible to fix this caching problem independently of whether backward compilation happens lazily vs. eagerly? (it feels to me like lazy backward compilation is around to stay, given that we are actively hardening compiled autograd)

(1) always add the forward compiled artifact to the cache eagerly when we get it
(2) only delay adding the backward compiled artifact into the cache to whenever we first generate it

I imagine this will require some work though, since IIRC we don't currently support this "partial" cache state (a cache entry where the fw has been populated but the bw has not)

**Comment by ezyang** [comment]
This bit me again today.
